### ISSUE TITLE
Juce 6/CMake fails to produce functioning *Non-Trivial* VST3 under Linux

### ISSUE THREAD

**Issue by braxtons12** [juce-framework/JUCE#718]
Running:
OS: Latest Version of Manjaro Linux
Compiler: Either of Clang 10 or GCC 9.3.0
IDE/Build Environment: VSCode with CMake Tools extension
Build System: The new CMake functionality

Trying to create anything non-trivial (eg, using files other than PluginEditor, PluginProcessor .h/.cpp, or user modules, etc) results in a "VST3 plugin" that is missing the plugin factory function
(project builds fine, but JUCE assertion failure in juce_VST3PluginFormat.cpp:878 and JUCE assertion failure in juce_VST3PluginFormat.cpp:1070 are triggered when trying to load in AudioPluginHost).

I've tried this with two different projects. 
One is an existing project developed under JUCE 5 that I simply created the CMakeLists.txt for to try and test the new functionality out. This builds and behaves as expect with the Makefile generated by Projucer v5, but has the mentioned behavior when using CMake.

The second is basically the demo project (just some renaming, and moving source files into "src" directory). The project compiles and functions properly when no extra functionality is added, but adding extra sources or user modules causes this issue (extra sources and user modules that were 100% fine w/ other projects under JUCE 5).

I have also noticed that I can have ZERO JUCE modules listed under target_link_libraries and everything will still compile and link properly... Not sure if this is intended or not or if it may be related.

I have not yet tested if this also occurs under Windows or MacOs.

Project CMakeLists.txt is attached.

[CMakeLists.txt](https://github.com/juce-framework/JUCE/files/4619778/CMakeLists.txt)

**Comment by braxtons12** [comment]
Hi, small update on this.

Tried a build with Projucer generated Makefiles this morning.
Through this I discovered that a teammate had refactored the sources in our user module without updating the root cpp file. This was causing some linker errors that for some reason weren't being caught by the build system when using CMake, but were caught when using Projucer generated Makefiles.

This only applies to the second mentioned project based on the demo project, however.
The first mentioned project uses no user modules and builds correctly via Projucer Makefiles, but incorrectly via CMake.

So it turns out there may be two issues at hand here.
1. Unknown ?? (mentioned project 1 case)
2. CMake system is setting up the linker commands incorrectly so failed linkage isn't failing the build (mentioned project 2 case)

**Comment by reuk** [comment]
It looks like the CMake build was omitting the `-Wl,--no-undefined` link flag, which meant that plugins with missing symbols would appear to build and link successfully. I'm fairly sure all of the issues you encountered were due to missing symbols in the final plugin. I'll push a change shortly to add the missing linker flag.

> I have also noticed that I can have ZERO JUCE modules listed under target_link_libraries and everything will still compile and link properly... Not sure if this is intended or not or if it may be related.

If your user modules specify other modules in their `dependencies` fields, all of the dependency modules will be included in the build automatically. That's why I was able to specify `juce_audio_processors` as the only dependency of `LibApex`, and link only against `LibApex` from the plugin target.
